butterfly-cli/build.gradle

```diff
@@ -12,20 +12,45 @@ dependencies {
             lib.commons_io,
             lib.log4j_api,
             lib.log4j_core,
-            lib.log4j_slf4j_impl,
             lib.disruptor
+    compile (lib.log4j_slf4j_impl) {
+        // favor newer 1.7.30
+        exclude(module: 'slf4j-api')
+    }
     compile (lib.spring_boot_starter) {
         exclude(module: 'commons-logging')
-        exclude(module:'logback-classic')
+        exclude(module: 'logback-classic')
+
+        // don't redirect log4j2 logs to slf4j because we bind slf4j to log4j2...
+        exclude(module: 'log4j-to-slf4j')
     }
     testCompile project(':butterfly-utilities')
     testCompile(lib.testng) {
         exclude(module: 'aopalliance')
         exclude(module: 'guava')
     }
     testCompile lib.mockito_all,
-            lib.powermock_module_testng,
-            lib.powermock_api_mockito
+                lib.hamcrest,
+                lib.javassist,
+                lib.objenesis
+    testCompile(lib.powermock_api_mockito) {
+        // favor version pulled in by lib.hamcrest
```

Comment:
These exclusions shouldn't be necessary. Gradle is supposed to take care of it. Also, not always the newer version is backward compatible, which may cause runtime issues.

Reply:
This was an attempt at fixing conflicting dependency versions. Gradle might "take care of it" but the fact remains that the app still pulls in conflicting dependencies. Without this exclusion, the app would pull in:

```
+--- org.powermock:powermock-api-mockito:1.6.5
 |    +--- org.mockito:mockito-core:1.10.19
 |    |    \--- org.hamcrest:hamcrest-core:1.1               <---
 |    +--- org.hamcrest:hamcrest-core:1.3
 |    \--- org.powermock:powermock-api-mockito-common:1.6.5
 |         +--- org.hamcrest:hamcrest-core:1.3               <---
```

... meaning the app has some deps that want hamcrest 1.1 and some deps that want hamcrest 1.3. This exclusion ensures a deliberate choice for which version of hamcrest the app should use.

&nbsp;
Also, I say attempt because I gave up half-way through. Although I fixed conflicts for compile-time deps, conflicts still remain for test-only deps, which require more effort to fix (and are arguably less important)

Reply:
I think there should be another pass or pull request to upgrade the test deps. There we could remove the manual exclusions like these.

Reply:
I see. Thanks for clarifying Myo!